database/rrdcontext.c

```diff
@@ -2057,6 +2105,481 @@ int rrdcontexts_to_json(RRDHOST *host, BUFFER *wb, time_t after, time_t before,
+
+    qt->db.minimum_latest_update_every = 0;
+    qt->db.first_time_t = 0;
+    qt->db.last_time_t = 0;
+
+    qt->id[0] = '\0';
+
+    qt->used = false;
+}
+void query_target_free(void) {
+    if(thread_query_target.used)
+        query_target_release(&thread_query_target);
+
+    freez(thread_query_target.query.array);
+    thread_query_target.query.array = NULL;
+    thread_query_target.query.size = 0;
+
+    freez(thread_query_target.metrics.array);
+    thread_query_target.metrics.array = NULL;
+    thread_query_target.metrics.size = 0;
+
+    freez(thread_query_target.instances.array);
+    thread_query_target.instances.array = NULL;
+    thread_query_target.instances.size = 0;
+
+    freez(thread_query_target.contexts.array);
+    thread_query_target.contexts.array = NULL;
+    thread_query_target.contexts.size = 0;
+
+    freez(thread_query_target.hosts.array);
+    thread_query_target.hosts.array = NULL;
+    thread_query_target.hosts.size = 0;
+}
+
+static void query_target_add_metric(QUERY_TARGET_LOCALS *qtl, RRDMETRIC_ACQUIRED *rma, RRDINSTANCE *ri, bool instance_matches_label_filters) {
+    QUERY_TARGET *qt = qtl->qt;
+
+    if(qt->metrics.used == qt->metrics.size) {
+        qt->metrics.size = (qt->metrics.size) ? qt->metrics.size * 2 : 1;
+        qt->metrics.array = reallocz(qt->metrics.array, qt->metrics.size * sizeof(RRDMETRIC_ACQUIRED *));
+    }
+    qt->metrics.array[qt->metrics.used++] = rrdmetric_acquired_dup(rma);
+
+    if(!instance_matches_label_filters)
+        return;
+
+    RRDMETRIC *rm = rrdmetric_acquired_value(rma);
+
+    rrdmetric_update_retention(rm);
+
+    bool timeframe_matches = (rm->first_time_t <= qt->window.before && rm->last_time_t >= qt->window.after) ? true : false;
+
+    if(timeframe_matches) {
+        RRDR_DIMENSION_FLAGS options = RRDR_DIMENSION_DEFAULT;
+
+        if (rrd_flag_check(rm, RRD_FLAG_HIDDEN)
+            || (rm->rrddim && rrddim_option_check(rm->rrddim, RRDDIM_OPTION_HIDDEN))) {
+            options |= RRDR_DIMENSION_HIDDEN;
+            options &= ~RRDR_DIMENSION_SELECTED;
+        }
+
+        if (qt->query.pattern) {
+            // we have a dimensions pattern
+            // lets see if this dimension is selected
+
+            if ((qtl->match_ids   && simple_pattern_matches(qt->query.pattern, string2str(rm->id)))
+             || (qtl->match_names && simple_pattern_matches(qt->query.pattern, string2str(rm->name)))
+                    ) {
+                // it matches the pattern
+                options |= (RRDR_DIMENSION_SELECTED | RRDR_DIMENSION_NONZERO);
+                options &= ~RRDR_DIMENSION_HIDDEN;
+            }
+            else {
+                // it does not match the pattern
+                options |= RRDR_DIMENSION_HIDDEN;
+                options &= ~RRDR_DIMENSION_SELECTED;
+            }
+        }
+        else {
+            // we don't have a dimensions pattern
+            // so this is a selected dimension
+            // if it is not hidden
```

Comment:
Now from Coverity:

```sh
*** CID 358670:  Control flow issues  (DEADCODE)
/database/rrdcontext.c: 2293 in query_target_add_metric()
2287                 // if it is not hidden
2288                 if(!(options & RRDR_DIMENSION_HIDDEN))
2289                     options |= RRDR_DIMENSION_SELECTED;
2290             }
2291     
2292             if((options & RRDR_DIMENSION_HIDDEN) && (options & RRDR_DIMENSION_SELECTED))
>>>     CID 358670:  Control flow issues  (DEADCODE)
>>>     Execution cannot reach this statement: "options &= 0xfffffffbU;".
2293                 options &= ~RRDR_DIMENSION_HIDDEN;
2294     
2295             if(!(options & RRDR_DIMENSION_HIDDEN) || (qt->request.options & RRDR_OPTION_PERCENTAGE)) {
2296                 // we have a non-hidden dimension
2297                 // let's add it to the query metrics
2298     
```

Reply:
I added this for clarity. With the current code this cannot be true, but I prefer it to be there because we are going to touch that part of the code later.

Reply:
Considering that I plan to run coverity again before we merge, I will mark this as resolved for while. :handshake: